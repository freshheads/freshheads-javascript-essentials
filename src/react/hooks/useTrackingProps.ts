import { Serializable } from '../../types/utility';

export type TrackingProps = {
    'data-tracking-event-name': string;
    'data-tracking-event-action': string;
    'data-tracking-event-context'?: Serializable;
};

/**
 * Applies uniform setup for tracking events, using attributes on DOM elements.
 * In Google Tag Manager these can be registered and used to, in turn, push
 * events to Google Analytics or other (tracking) platforms.
 *
 * @param {string} eventName        i.e. "preorder"
 * @param {string} eventAction      i.e. "submit"
 * @param {Serializable=} context   i.e. "{ subscribeToNewsletter: false }"
 */
export default function useTrackingProps(
    eventName: string,
    eventAction: string,
    context?: Serializable
): TrackingProps {
    let props: TrackingProps = {
        'data-tracking-event-name': eventName,
        'data-tracking-event-action': eventAction,
    };

    if (context) {
        props['data-tracking-event-context'] = JSON.stringify(context);
    }

    return props;
}
