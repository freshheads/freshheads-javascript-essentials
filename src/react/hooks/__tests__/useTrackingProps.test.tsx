import { render } from '@testing-library/react';
import useTrackingProps from '../useTrackingProps';

import React from 'react';

describe('useTrackingProps()', () => {
    describe('When applied on a JSX element', () => {
        describe('..with context supplied', () => {
            it('should render the tracking props on it correctly', () => {
                const trackingProps = useTrackingProps('preorder', 'submit', {
                    subscribeToNewsletter: true,
                });

                const { asFragment } = render(<button {...trackingProps} />);

                expect(asFragment()).toMatchSnapshot();
            });
        });

        describe('..without context supplied', () => {
            it('should render the tracking props on it correctly', () => {
                const trackingProps = useTrackingProps('preorder', 'submit');

                const { asFragment } = render(<button {...trackingProps} />);

                expect(asFragment()).toMatchSnapshot();
            });
        });
    });
});
