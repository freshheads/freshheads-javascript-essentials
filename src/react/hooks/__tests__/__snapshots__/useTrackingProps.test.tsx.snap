// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useTrackingProps() When applied on a JSX element ..with context supplied should render the tracking props on it correctly 1`] = `
<DocumentFragment>
  <button
    data-tracking-event-action="submit"
    data-tracking-event-context="{\\"subscribeToNewsletter\\":true}"
    data-tracking-event-name="preorder"
  />
</DocumentFragment>
`;

exports[`useTrackingProps() When applied on a JSX element ..without context supplied should render the tracking props on it correctly 1`] = `
<DocumentFragment>
  <button
    data-tracking-event-action="submit"
    data-tracking-event-name="preorder"
  />
</DocumentFragment>
`;
